src/consensus/validation.h

```diff
@@ -137,8 +137,8 @@ class CValidationState {
     }
     ValidationInvalidReason GetReason() const { return m_reason; }
     unsigned int GetRejectCode() const { return chRejectCode; }
-    std::string GetRejectReason() const { return strRejectReason; }
-    std::string GetDebugMessage() const { return strDebugMessage; }
+    const std::string& GetRejectReason() const { return strRejectReason; }
+    const std::string& GetDebugMessage() const { return strDebugMessage; }
```

Comment:
Why this change? :)

Reply:
See https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543330180.

@jnewbery Regarding my previous comment, I'm also happy to remove this commit if that's preferred. Seems I had not read your comment closely enough about it not being performance critical. Though note my concerns about macro vs function logging.

Reply:
What is the object lifetime `strRejectReason` and `strDebugMessage`?

Are they guaranteed to outlive the callers of `GetRejectReason` or `GetDebugMessage`?

Reply:
Their lifetimes are that of the corresponding `CValidationState` as mentioned in https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-527575841. All current uses are guaranteed to outlive the callers as they either make a copy or are passed by const reference to other functions that do something similar. There's no such guarantee for future callers though.

Reply:
I don't think 6c730c3926d1871d8ce4c63eb6752808d7c3484a is a good idea: it introduces a very sharp edge in consensus critical code which is not offset by any measurable gain AFAICT.

We should always err on the side of caution when touching consensus critical code: if there is a trade-off between safety (current or future) and speed, then safety is generally much more important. In the cases where speed is critical for some reasons we should always measure before optimising :)

Somewhat related: https://www.youtube.com/watch?v=3MB2iiCkGxg&t=870

Reply:
Fair enough. I've removed 6c730c3.